syntax = "proto3";

package MailTaxiAPI;

message Coordinate{
    double latitude = 1;
    double longitude = 2;
}

message LocationData{
    string token = 1;
    Coordinate coordinate = 2;
    bool isDriver = 3;
}

message DriverLocationData{
    Coordinate coordinate = 1;
    bool   busy = 2;
}

enum OrderStatus{
    economy = 0;
    business = 1;
}

message Driver{
    OrderStatus status = 1;
    bool        children = 2;
    bool        animals = 3;
}

message DriverInformation{
	string token = 1;
	Driver driver = 2;
	Coordinate coordinate = 3;
}


service OrderService{
    rpc GetDrivers(GetDriversRequest) returns (GetDriversResponse){}
    rpc RegDriver(RegDriverRequest) returns (RegDriverResponse){}
    rpc NewOrderForDriver(NewOrderForDriverRequest)  returns (NewOrderForDriverResponse){}
}
message OrderData{
    Coordinate ClientAddress = 1;
    Coordinate destinationAddress = 2;
}

message NewOrderForDriverRequest{
    string token = 1;
    OrderData orderData = 2;
}

message NewOrderForDriverResponse{
    bool success = 1;
}

message GetDriversRequest{
    Coordinate coordinate = 1;
    OrderStatus status = 2;
    bool        children = 3;
    bool        animals = 4;
}


message GetDriversResponse{
    repeated DriverInformation information = 1;
}

message RegDriverRequest{
  Driver driver = 1;
  Coordinate coordinate = 2;
  string token = 3  ;
}

message RegDriverResponse{
  bool success = 1;
}


service DistributionOfOrders{
    rpc MakeOrder(MakeOrderRequest) returns (MakeOrderResponse){}
    rpc UserAuthorization(UserAuthorizationRequest) returns (UserAuthorizationResponse){}
    rpc StartChangingDriver(StartChangingDriverRequest) returns (StartChangingDriverResponse){}
    rpc RegistrationUser(RegistrationUserRequest) returns (RegDriverResponse){}
}

message RegistrationUserRequest{
    string login = 1;
    string password = 2;
    bool isDriver = 3;
    Driver driver = 4;
}
message RegistrationUserResponse{
    string token = 1;
}

message MakeOrderRequest{
    OrderData  orderData = 1;
    OrderStatus type = 2;
    bool        children = 3;
    bool        animals = 4;
    string token = 5;
}

message MakeOrderResponse{
    string driverToken = 1;
}

message StartChangingDriverRequest{
    string token = 1;
    Coordinate coordinate = 2;
}

message StartChangingDriverResponse{
    bool success = 1;
}

message UserAuthorizationRequest{
    string login = 1;
    string password = 2;
}

message UserAuthorizationResponse{
    string token = 1;
}